File: 2/2/pdoo/prac/ruby/gameuniverse.rb
#encoding:utf-8

require "./lib/GameUniverseToUI.rb"
require "./lib/GameStateController.rb"
require_relative "./dice.rb"
require_relative "./spacestation.rb"
require_relative "./enemystarship"
require_relative "./gamecharacter.rb"
require_relative "./combatresult.rb"

module Deepspace
  # Clase que desarrolla la partida
  class GameUniverse
    # Puntos necesarios para ganar
    @@WIN = 10
    
    attr_reader :gameState

    def initialize
      @gameState = GameStateController.new
      @turns = 0
      @dice = Dice.new
      @currentStation = nil
      @currentEnemy = nil
      @spaceStations = Array.new
    end

    # Implementación en la práctica 3
    def combatGo(station,enemy)
    end

    # Implementación en la práctica 3
    def combat
    end
    
    # La nave del turno descarta el hangar
    def discardHangar
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardHangar
      end
    end

    # La nave del turno descarta un potenciador de escudo
    def discardShieldBooster(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardShieldBooster(i)
      end
    end

    # La nave del turno descarta un potenciador de escudo de su hangar
    def discardShieldBoosterInHangar(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardShieldBoosterInHangar(i)
      end
    end

    # La nave del turno descarta un arma
    def discardWeapon(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardWeapon(i)
      end
    end

    # La nave del turno descarta un arma de su hangar
    def discardWeaponInHangar(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardWeaponInHangar(i)
      end
    end

    def getUIVersion
      GameUniverseToUI.new(self)
    end

    # Comprueba si la nave del turno ha ganado la partida
    def haveAWinner
      if @currentStation.nMedals>=@@WIN
        return true
      end
      return false
    end

    # Implementación en la práctica 3
    def init(names)
    end

    # Monta un escudo con índice 'i' en la nave
    def mountShieldBooster(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.mountShieldBooster(i)
      end
    end

    # Monta un arma con índice 'i' en la nave
    def mountWeapon(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.mountWeapon(i)
      end
    end

    # Implementación en la práctica 3
    def nextTurn
    end

    def to_s
      getUIVersion.to_s
    end
  end
end

File: 2/2/pdoo/prac/java/deepspace/src/Deepspace/GameUniverse.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.ArrayList;

/**
 * @brief Clase que controla la partida
 */
class GameUniverse {
    private final static int WIN = 10;
    
    private ArrayList<SpaceStation> spaceStations = new ArrayList<>();
    private SpaceStation currentStation;
    private EnemyStarShip currentEnemy;
    private GameStateController gameState;
    private Dice dice;
    private int currentSpaceStationIndex;
    private int turns;
    
    GameUniverse() {
        gameState = new GameStateController();
        dice = new Dice();
        turns = 0;
    }
    
    CombatResult combat(SpaceStation station,EnemyStarShip enemy) {
        throw new UnsupportedOperationException();
    }
    
    public CombatResult combat() {
        throw new UnsupportedOperationException();
    }
    
    public void discardHangar() {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardHangar();
    }
    
    public void discardShieldBooster(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardShieldBooster(i);
    }
    
    public void discardShieldBoosterInHangar(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardShieldBoosterInHangar(i);
    }
    
    public void discardWeapon(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardWeapon(i);
    }
    
    public void discardWeaponInHangar(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardWeaponInHangar(i);
    }
    
    public GameState getState() { return gameState.getState(); }
    
    GameUniverseToUI getUIversion() { return new GameUniverseToUI(currentStation,currentEnemy); }
    
    public boolean haveAWinner() { return currentStation.getNMedals()==WIN; }
    
    public void init(ArrayList<String> names) {
        throw new UnsupportedOperationException();
    }
    
    public void mountShieldBooster(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.mountShieldBooster(i);
    }
    
    public void mountWeapon(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.mountWeapon(i);
    }
    
    public boolean nextTurn() {
        throw new UnsupportedOperationException();
    }
    
    @Override
    public String toString() {
        String aux = "Estación espacial actual:\n"+currentStation.toString();
        aux += "\nEnemigo actual:\n"+currentEnemy.toString();
        return aux;
    }
}
